Fix accumulated difficulty while syncing

stringhandler opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Fix a bug where the accumulated difficulty was not set in the chain metadata db while syncing blocks. Also moved the `target_difficulty` fields around in the GRPC proto file and added `reward` and `total_fees` to make it easier for miners

Commit:
Fix accumulated difficulty while syncing

hansieodendaal commented:
Some issues uncovered while testing, maybe not all related to this PR:

**1.** Initial sync logic seems broken, either achieved early with a new sync, or not at all if starting with a blockchain at the tip. The latter means merge mining will not start, so still trying to test if the target difficulty gets passed through ok.

**2.** Seeing some RPC response timeout errors with initial block sync:
```
[c::bn::block_sync] DEBUG Block sync failed: RPC request failed: Timeout: Response timed out
[c::bn::base_node] TRACE Base Node event in State [Synchronizing blocks]:  Block Synchronization Failed
...
[c::bn::block_sync] DEBUG Block sync failed: RPC error: RPC negotiation timed out
[c::bn::base_node] TRACE Base Node event in State [Synchronizing blocks]:  Block Synchronization Failed
```
**3.** With this base node having `19979` headers and `92` blocks in its db, another peer requested `19979` blocks, but the base node could only provide `92`, ending in an error:
```
2021-01-06 16:14:31.569714700 [c::base_node::sync_rpc] DEBUG Initiating block sync with peer `1cc148eeaa18107edfe665a06b` from height 1 to 19979
...
2021-01-06 16:14:31.626960700 [c::base_node::sync_rpc] ERROR Internal error: The requested Block was not found via end height:96 in the database
2021-01-06 16:14:31.627000000 [c::base_node::sync_rpc] DEBUG Block sync round complete for peer `1cc148eeaa18107edfe665a06b`.
```
**4.** I see accumulated difficulty being calculated correctly when the block is added to the db, ~~but not being transmitted correctly to other nodes. For `ece4c616a156c1e4fe34d94d8c`~~
```
[c::bn::block_sync] DEBUG Block body #93 added in 43ms, Tot_acc_diff 1,794,638,658,122, Monero 1, SHA3 1,794,638,658,122
```
~~```~~
~~[c::bn::chain_state_sync_service] DEBUG Received chain metadata from NodeId 'ece4c616a156c1e4fe34d94d8c' #93, Acc_diff 1~~
~~```~~

Commit:
Fix initial sync logic error, add acc_diff logs
Before RPC block sync, initial sync was triggered when the
set of requested blocks were added. This test is not robust
anymore and is therefore removed. Added acc_diff log entries.

hansieodendaal commented:
I proposed some changes in https://github.com/mikethetike/tari/pull/3 to fix one of the initial sync logic errors, as well as some changes to logs to also print accumulated difficulty.

Target difficulty to the merge mining proxy still to be tested with `wait_for_initial_sync_at_startup = false` flag in config due to problems detecting initial sync, will revert.

I think after that we can merge this PR, as the issues picked up are unrelated.

stringhandler committed:
Merge pull request #3 from hansieodendaal/mb-fix-accum-diff-ho-test

Fix initial sync logic error, add acc_diff logs

hansieodendaal commented:
Target difficulty to merge mining proxy now constant at `60,000`, still a problem. At the base node:
``` rust
2021-01-07 10:19:03.659979900 [c::bn::comms_interface::inbound_handler] TRACE Calculating target difficulty at height: 19989 for PoW: Monero
2021-01-07 10:19:03.663072200 [c::bn::comms_interface::inbound_handler] DEBUG Target difficulty 60,000 for PoW Monero
```
At the merge mining proxy:
``` rust
proxy\proxy.log:
2021-01-07 10:03:03.424419400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(460806), Selected(60000) 
2021-01-07 10:06:17.861219500 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(461022), Selected(60000) 
2021-01-07 10:07:26.216547100 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(460194), Selected(60000) 
2021-01-07 10:11:48.485448400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(460030), Selected(60000) 
2021-01-07 10:14:18.681844000 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(458514), Selected(60000) 
2021-01-07 10:15:53.533766900 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(460651), Selected(60000) 
2021-01-07 10:16:19.575583300 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(461057), Selected(60000) 
2021-01-07 10:18:09.719089800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(459859), Selected(60000) 
2021-01-07 10:18:31.870351000 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(460564), Selected(60000) 
```

hansieodendaal commented:
Also picking up this issue with `monerod`:
``` rust
2021-01-07 10:06:29.388851600 [tari_mm_proxy::xmrig] DEBUG Proxying request: POST http://18.133.55.120:38081/json_rpc "submitblock" 
2021-01-07 10:06:29.928961900 [tari_mm_proxy::xmrig] DEBUG Block submitted to monerod, response is: {"error":{"code":-6,"message":"Wrong block blob"},"id":21,"jsonrpc":"2.0"} 
2021-01-07 10:06:29.929024100 [tari_mm_proxy::xmrig] ERROR Error handling request: Missing data:blocktemplate blob invalid 
```
Maybe a different issue.

hansieodendaal commented:
It turns out up to XMRig 6.6.2 work with our latest dev release, but XMRig 6.7.0 does not; the cause of the `monerod` problem noted above^.

hansieodendaal commented:
Target difficulties working properly, once a compatible XMRig version was sorted out. At the proxy:
``` rust
   7510: 2021-01-08 13:21:19.750284500 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503685), Selected(60000) 
   7730: 2021-01-08 13:21:40.414292600 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503685), Selected(60000) 
   7830: 2021-01-08 13:21:48.624695600 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503685), Selected(60000) 
   8020: 2021-01-08 13:22:05.674238200 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503434), Selected(60000) 
   8200: 2021-01-08 13:22:21.725925200 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503434), Selected(60000) 
   8650: 2021-01-08 13:23:05.204353000 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503434), Selected(60000) 
   8760: 2021-01-08 13:23:14.897669400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(503434), Selected(60000) 
   8950: 2021-01-08 13:23:29.948728400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60000), Monero(505430), Selected(60000) 
   9080: 2021-01-08 13:23:40.437110400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (60197), Monero(505430), Selected(60197) 
   9217: 2021-01-08 13:23:45.531595800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (62236), Monero(505353), Selected(62236) 
   9255: 2021-01-08 13:23:45.533336300 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (62236), Monero(505353), Selected(62236) 
   9350: 2021-01-08 13:23:50.467847700 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (62236), Monero(505915), Selected(62236) 
   9480: 2021-01-08 13:24:00.336077600 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (64425), Monero(505915), Selected(64425) 
   9570: 2021-01-08 13:24:05.308487800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (64425), Monero(507011), Selected(64425) 
   9810: 2021-01-08 13:24:26.420230700 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (66712), Monero(507011), Selected(66712) 
   9910: 2021-01-08 13:24:35.113641700 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (69037), Monero(507011), Selected(69037) 
  10550: 2021-01-08 13:25:37.649643800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (71564), Monero(507011), Selected(71564) 
  10700: 2021-01-08 13:25:51.339678800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (73659), Monero(507011), Selected(73659) 
  10950: 2021-01-08 13:26:14.258181800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (76392), Monero(507011), Selected(76392) 
  11260: 2021-01-08 13:26:41.479341800 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (76392), Monero(506505), Selected(76392) 
  11600: 2021-01-08 13:27:12.668914100 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (78887), Monero(506505), Selected(78887) 
  11720: 2021-01-08 13:27:22.728267600 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (81766), Monero(506505), Selected(81766) 
  11840: 2021-01-08 13:27:30.673317500 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (85124), Monero(506505), Selected(85124) 
  12010: 2021-01-08 13:27:46.113676500 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (88737), Monero(506505), Selected(88737) 
  12400: 2021-01-08 13:28:21.341036100 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (88737), Monero(505940), Selected(88737) 
  12600: 2021-01-08 13:28:38.205553600 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (91994), Monero(505940), Selected(91994) 
  12695: 2021-01-08 13:28:39.665497300 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (96001), Monero(504964), Selected(96001) 
  12740: 2021-01-08 13:28:40.162667400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (96001), Monero(504964), Selected(96001) 
  13170: 2021-01-08 13:29:20.647456400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (100604), Monero(504964), Selected(100604) 
  13355: 2021-01-08 13:29:31.635804400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (104883), Monero(503650), Selected(104883) 
  13400: 2021-01-08 13:29:31.678354300 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (104883), Monero(503650), Selected(104883) 
  13810: 2021-01-08 13:30:07.979857400 [tari_mm_proxy::xmrig] INFO  Difficulties: Tari (104883), Monero(504758), Selected(104883)

hansieodendaal commented:
GRPC calls submitting Monero merge mined blocks to the base node returning really quickly now, so PR #2483 also did its job! See the first block submission and validation taking ~10 times longer due to new RandomX VM being created.
``` rust
proxy\proxy.log:
2021-01-08 13:10:37.542192400 [tari_mm_proxy::xmrig] DEBUG Submitted block #20031 to Tari node in 440ms (SubmitBlock) 
2021-01-08 13:10:54.154025800 [tari_mm_proxy::xmrig] DEBUG Submitted block #20032 to Tari node in 39ms (SubmitBlock) 
2021-01-08 13:11:27.672252200 [tari_mm_proxy::xmrig] DEBUG Submitted block #20033 to Tari node in 40ms (SubmitBlock) 
2021-01-08 13:11:53.798286400 [tari_mm_proxy::xmrig] DEBUG Submitted block #20034 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:12:01.906757700 [tari_mm_proxy::xmrig] DEBUG Submitted block #20035 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:12:12.837098500 [tari_mm_proxy::xmrig] DEBUG Submitted block #20036 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:13:41.424806200 [tari_mm_proxy::xmrig] DEBUG Submitted block #20037 to Tari node in 43ms (SubmitBlock) 
2021-01-08 13:13:55.580795000 [tari_mm_proxy::xmrig] DEBUG Submitted block #20038 to Tari node in 36ms (SubmitBlock) 
2021-01-08 13:14:34.335980600 [tari_mm_proxy::xmrig] DEBUG Submitted block #20039 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:14:54.887449500 [tari_mm_proxy::xmrig] DEBUG Submitted block #20040 to Tari node in 36ms (SubmitBlock) 
2021-01-08 13:15:03.382536200 [tari_mm_proxy::xmrig] DEBUG Submitted block #20041 to Tari node in 37ms (SubmitBlock) 
2021-01-08 13:15:16.877458400 [tari_mm_proxy::xmrig] DEBUG Submitted block #20042 to Tari node in 80ms (SubmitBlock) 
2021-01-08 13:15:27.253998700 [tari_mm_proxy::xmrig] DEBUG Submitted block #20043 to Tari node in 60ms (SubmitBlock) 
2021-01-08 13:15:40.761403100 [tari_mm_proxy::xmrig] DEBUG Submitted block #20044 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:16:13.127671400 [tari_mm_proxy::xmrig] DEBUG Submitted block #20045 to Tari node in 79ms (SubmitBlock) 
2021-01-08 13:16:23.932511500 [tari_mm_proxy::xmrig] DEBUG Submitted block #20046 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:16:32.121736600 [tari_mm_proxy::xmrig] DEBUG Submitted block #20047 to Tari node in 39ms (SubmitBlock) 
2021-01-08 13:17:37.263852200 [tari_mm_proxy::xmrig] DEBUG Submitted block #20048 to Tari node in 39ms (SubmitBlock) 
2021-01-08 13:17:51.333679000 [tari_mm_proxy::xmrig] DEBUG Submitted block #20049 to Tari node in 37ms (SubmitBlock) 
2021-01-08 13:18:02.818628700 [tari_mm_proxy::xmrig] DEBUG Submitted block #20050 to Tari node in 39ms (SubmitBlock) 
2021-01-08 13:18:29.846612400 [tari_mm_proxy::xmrig] DEBUG Submitted block #20051 to Tari node in 62ms (SubmitBlock) 
2021-01-08 13:18:36.515388300 [tari_mm_proxy::xmrig] DEBUG Submitted block #20052 to Tari node in 39ms (SubmitBlock) 
2021-01-08 13:19:22.147802200 [tari_mm_proxy::xmrig] DEBUG Submitted block #20053 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:19:43.797834800 [tari_mm_proxy::xmrig] DEBUG Submitted block #20054 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:20:33.934532300 [tari_mm_proxy::xmrig] DEBUG Submitted block #20055 to Tari node in 35ms (SubmitBlock) 
2021-01-08 13:20:50.445470000 [tari_mm_proxy::xmrig] DEBUG Submitted block #20056 to Tari node in 36ms (SubmitBlock) 
2021-01-08 13:21:14.148268300 [tari_mm_proxy::xmrig] DEBUG Submitted block #20057 to Tari node in 37ms (SubmitBlock) 
2021-01-08 13:21:34.772867400 [tari_mm_proxy::xmrig] DEBUG Submitted block #20058 to Tari node in 37ms (SubmitBlock) 
2021-01-08 13:21:42.982502300 [tari_mm_proxy::xmrig] DEBUG Submitted block #20059 to Tari node in 38ms (SubmitBlock) 
2021-01-08 13:22:00.053990700 [tari_mm_proxy::xmrig] DEBUG Submitted block #20060 to Tari node in 39ms (SubmitBlock) 
2021-01-08 13:22:16.111470600 [tari_mm_proxy::xmrig] DEBUG Submitted block #20061 to Tari node in 38ms (SubmitBlock) 
```

hansieodendaal commented:
Well done @mikethetike!

hansieodendaal reviewed (approved):
Tested (a) accumulated difficulty and (b) target difficulty to proxy fixes, all good.

LGTM!
